File: python/src/siphon/base64_stream.py
"""Chunked Base64 encoding so inline fallback never buffers a whole file."""

from __future__ import annotations

import base64
from typing import Iterator


def encode_chunks_to_base64(chunks: Iterator[bytes]) -> str:
    parts: list[str] = []
    remainder = b""

    for chunk in chunks:
        buffer = remainder + chunk
        usable_len = (len(buffer) // 3) * 3
        usable, remainder = buffer[:usable_len], buffer[usable_len:]
        if usable:
            parts.append(base64.b64encode(usable).decode("ascii"))

    if remainder:
        parts.append(base64.b64encode(remainder).decode("ascii"))

    return "".join(parts)

File: typescript/src/base64Stream.ts
export async function encodeChunksToBase64(chunks: AsyncIterable<Buffer>): Promise<string> {
  const parts: string[] = [];
  let remainder = Buffer.alloc(0);

  for await (const chunk of chunks) {
    const buffer = Buffer.concat([remainder, chunk]);
    const usableLength = Math.floor(buffer.length / 3) * 3;
    const usable = buffer.subarray(0, usableLength);
    remainder = buffer.subarray(usableLength);
    if (usable.length > 0) {
      parts.push(usable.toString("base64"));
    }
  }

  if (remainder.length > 0) {
    parts.push(remainder.toString("base64"));
  }

  return parts.join("");
}
